### PR TITLE
Resolves issue #936

### PR DESCRIPTION
**Pull request by lependu**
The original discussion is #936.

#895 made possible to accept request without body and with valid `Content-Type` header **other than** `application/json`.

This change makes possible to do the same if the request method is `DELETE` and the `Content-Type` header is `application/json`.

The PR also separates how we handle the `OPTIONS` and `DELETE` methods based on https://github.com/fastify/fastify/issues/936#issuecomment-397460128.

Best of my knowledge these two RFC paragraph are the closest to our use case, but none of them regulates it explicitly:
*From [RFC 7231 Section 3.1.1.5](https://tools.ietf.org/html/rfc7231#section-3.1.1.5)*
> A sender that generates a message containing a payload body SHOULD
> generate a Content-Type header field in that message unless the
> intended media type of the enclosed representation is unknown to the sender.

*From [RFC  7231 Section 4.3.5](https://tools.ietf.org/html/rfc7231#section-4.3.5)*
> A payload within a DELETE request message has no defined semantics;
> sending a payload body on a DELETE request might cause some existing
> implementations to reject the request.
</details>

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/handleRequest.js [hunk 1]
```diff
@@ -37,7 +37,28 @@ function handleRequest (req, res, params, context) {
     return
   }
 
-  if (method === 'OPTIONS' || method === 'DELETE') {
+  if (method === 'DELETE') {
+    if (contentType === undefined) {
+      handler(reply)
+    } else {
+      // https://github.com/fastify/fastify/issues/936
```
Comment (StarpTech) [inline lib/handleRequest.js]: We have never added such comments. I think we can remove it.

File: lib/handleRequest.js [hunk 2]
```diff
@@ -37,7 +37,28 @@ function handleRequest (req, res, params, context) {
...
+      if (
+        contentType === 'application/json' &&
```
Comment (StarpTech) [inline lib/handleRequest.js]: Use `contentType.indexOf('application/json')` as in https://github.com/fastify/fastify/blob/e51ad47e437a7e6ecdbe65cc18cfd2f3fa57d3dc/lib/reply.js#L74 in order to be consistent

**Comment by lependu** [comment]
@delvedor:
It started with https://github.com/fastify/fastify/issues/894#issue-316256392 which solved by 895, with a clear statement in https://github.com/fastify/fastify/pull/895#issuecomment-383156467 which says that requests with `DELETE` method with `application/json` `Content-Type` header - same as any other invalid JSON payload - will still throw.

Then came this issue with the use case described in https://github.com/fastify/fastify/issues/936#issue-322680784.

Because both cited issue had a `bug` label and the current one had a `help wanted` label too, I had a false impression that You want to extend the support with handling request with `DELETE` method and `application/json` `Content-Type` header and no body. Now I had a sad feeling that I should asked first (as I always did so far except this case). I am really sorry for that.

The code what @nwoltman suggests in his previous comment would:
- still throw `400 Bad Request` error with `DELETE` method and `application/json` `Content-Type` header and no body, because of [JSON.parse()](https://github.com/fastify/fastify/blob/4949a9c53322eeb9ba6e26f4e2c367d79745ef6a/lib/ContentTypeParser.js#L161-L166). That function is also the main reason of the complexity, and the reason why we need to separate the `OPTIONS` and `DELETE` methods (also see below), and check if the `Content-Type` header is `application/json` in the case of the `DELETE` method.
- start to throw `400 Bad Request` errors in case of `OPTIONS` method.
- break [request should be defined in onSend Hook on options request with content type application/x-www-form-urlencoded](https://github.com/fastify/fastify/blob/4949a9c53322eeb9ba6e26f4e2c367d79745ef6a/test/internals/handleRequest.test.js#L161) functionality.

**Comment by nwoltman** [comment]
@lependu Here are my responses to your bullet points on the code I proposed in my [previous comment](https://github.com/fastify/fastify/pull/1027#issuecomment-403256866):

- It would only throw a `400 Bad Request` on `DELETE` requests with a `Content-Type: application/json` if there is also a `Content-Length` or `Transfer-Encoding` header. If there is no `Content-Length/Transfer-Encoding` header, body parsing will be skipped, so it won't throw. However, if a `DELETE` request sends a `Content-Type: application/json` header and a `Content-Length: 0` header, the request is indicating that it has a zero-length JSON body. This is an invalid request because the empty string `''` is invalid JSON, so `400 Bad Request` is the appropriate response.
- Unless I'm missing something, `OPTIONS` requests shouldn't start to throw `400 Bad Request` errors. In fact, it should throw those **less often**. The previous code would try to parse the body of an `OPTIONS` request only if it had a `Content-Type` header. My suggested code requires both a `Content-Type` header and a `Content-Length/Transfer-Encoding` header, which means it will try to parse bodies less often, and therefore throw less often.
- If my suggested code gets implemented, it would make sense for that test to change. It would respond with a `200 OK` because a body isn't being sent with the request (so body parsing would be skipped).

File: test/helper.js [hunk 3]
```diff
@@ -236,16 +236,18 @@ module.exports.payloadMethod = function (method, t) {
...
-      }, (err, response, body) => {
-        t.error(err)
-        t.strictEqual(response.statusCode, 400)
-      })
+      if (upMethod !== 'OPTIONS') {
+        sget({
+          method: upMethod,
+          url: 'http://localhost:' + fastify.server.address().port,
+          body: '',
+          headers: { 'Content-Type': 'application/json' },
```
Comment (nwoltman) [inline test/helper.js]: To avoid the separate case for the OPTIONS request, you could remove the `body: '',` line and add a `'Content-Length': '0'` header.

File: test/helper.js [hunk 4]
```diff
@@ -239,8 +239,10 @@ module.exports.payloadMethod = function (method, t) {
       sget({
         method: upMethod,
         url: 'http://localhost:' + fastify.server.address().port,
-        body: '',
-        headers: { 'Content-Type': 'application/json' },
+        headers: {
```
Comment (delvedor) [inline test/helper.js]: I think you should add another test here instead of update this.

File: test/internals/handleRequest.test.js [hunk 5]
```diff
@@ -183,8 +183,8 @@ test('request should be defined in onSend Hook on options request with content t
       }
     }, (err, response, body) => {
       t.error(err)
-      // a 415 error is expected because of missing content type parser
-      t.strictEqual(response.statusCode, 415)
+      // Body parsing skipped, so no body sent
+      t.strictEqual(response.statusCode, 200)
```
Comment (delvedor) [inline test/internals/handleRequest.test.js]: Is this a breaking change?

File: test/custom-parser.test.js [hunk 6]
```diff
@@ -647,7 +647,7 @@ test('Should parse empty bodies as a string', t => {
   })
 
   fastify.route({
-    method: ['POST', 'DELETE'],
+    method: ['POST'],
```
Comment (delvedor) [inline test/custom-parser.test.js]: I would expect this to work as it was before. cc @nwoltman

File: lib/handleRequest.js [hunk 7]
```diff
@@ -38,10 +38,16 @@ function handleRequest (req, res, params, context) {
...
+      contentType !== undefined &&
+      (
+        headers['transfer-encoding'] !== undefined ||
+        headers['content-length'] !== undefined
+      )
+    ) {
       context.contentTypeParser.run(contentType, handler, request, reply)
+    } else {
+      handler(reply)
     }
```
Comment (mcollina) [inline lib/handleRequest.js]: Why can't we merge this block with the one for POST?

File: test/helper.js (hunk 4 above)
Comment (lependu) [reply test/helper.js]: In case of `OPTIONS` method the body parsing would be skipped.
May I restore the tests of other methods which I removed because of https://github.com/fastify/fastify/pull/1027#pullrequestreview-135274087?

File: test/custom-parser.test.js (hunk 6 above)
Comment (lependu) [reply test/custom-parser.test.js]: It will skip the body parsing too. May I add a new test which cheks if the body is null?

Comment (nwoltman) [reply test/custom-parser.test.js]: This test is a little misleading because `body: '',` gets passed as an option to `simple-get` but `simple-get` ignores this and doesn't set the `Content-Length` header. With the new changes, `DELETE` requests are interpreted as there is no body so the body won't be parsed (whereas `POST` requests expect a body so it will still be parsed).

This test could stay mostly the same if `Content-Length: 0` were added to the `headers` passed to `sget`.

File: lib/handleRequest.js (hunk 7 above)
Comment (nwoltman) [reply lib/handleRequest.js]: The logic is a little bit different between the types of requests.

The POST request logic will always parse the request body if a `Content-Type` is present, so if we used this same logic for DELETE requests, #936 wouldn't get fixed. The POST request logic will also try to parse the body when a `Content-Type` isn't present, which goes against the requirements for OPTIONS requests.

File: test/internals/handleRequest.test.js (hunk 5 above)
Comment (nwoltman) [reply test/internals/handleRequest.test.js]: To be safe we could consider it a breaking change, but I don't think it is. There isn't a body being sent with the request so I'd consider the old `415 Unsupported Media Type` response to be a bug.

Comment (delvedor) [reply test/internals/handleRequest.test.js]: I agree. @mcollina?

File: lib/handleRequest.js (hunk 2 above)
Comment (lependu) [reply lib/handleRequest.js]: @StarpTech Can you review it again, please?

**Comment by lependu** [comment]
@StarpTech blocking this because of https://github.com/fastify/fastify/pull/1027#discussion_r200820622 which was updated.
Is https://github.com/fastify/fastify/pull/1027#discussion_r201025329 is still on the table, @delvedor?

File: test/internals/handleRequest.test.js (hunk 5 above)
Comment (mcollina) [reply test/internals/handleRequest.test.js]: Agreed
